nixos/nsswitch cleanup nss modules

flokli opened the pull request:
This moves the remaining NSS modules to their corresponding modules, and contains some follow-up fixes, where we missed adding to the `group` database (as the `passwdArray` was used both for `system.nssDatabases.passed` and `system.nssDatabases.group`).

It also removes the conditional loading of systemd nss modules, as requested in https://github.com/NixOS/nixpkgs/pull/86940#discussion_r420122114.

closes #86350.

###### Motivation for this change
#86350

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

flokli committed:
nixos/sssd: add to system.nssDatabases.group too

nixos/modules/config/nsswitch.nix uses `passwdArray` for both `passwd`
and `group`, but when moving this into the sss module in
edddc7c82a848f1b1f23344a338b5d56b38178c0, it didn't get split
appropriately.

flokli committed:
nixos/google-oslogin: add to system.nssDatabases.group too

nixos/modules/config/nsswitch.nix uses `passwdArray` for both `passwd`
and `group`, but when moving this into the google-oslogin module in
4b71b6f8fa5f7b06ea5d0dd37d44a73757c7f142, it didn't get split
appropriately.

flokli committed:
nixos/systemd: add to system.nssDatabases.group too

nixos/modules/config/nsswitch.nix uses `passwdArray` for both `passwd`
and `group`, but when moving this into the systemd module in
c0995d22eed1a19ac9442c8460c18dd6a4c389b7, it didn't get split
appropriately.

flokli committed:
nixos/ldap: move nss database configuration into ldap module

now that passwdArray and shadowArray aren't used anymore, these can be
folded.

flokli committed:
nixos/avahi: move nss database configuration into avahi module

flokli committed:
nixos/samba: move nss database configuration into samba module

flokli committed:
nixos/sssd: drop assertion

This is now already triggered by the nsswitch module, as we set
system.nssModules.

flokli committed:
nixos/systemd: enable systemd-provided nss modules unconditionally

A disabled nscd breaks nss module loading on NixOS, and systemd without
its nss modules doesn't really work either - instead of silently
disabling its nss modules if nscd is disabled, let the assertion in
nsswitch handle this.

flokli committed:
nixos/nsswitch: update comment next to assertion

dasJ commented:
@GrahamcOfBorg test google-oslogin avahi

Samba test doesn't seem to be testing the right stuff.
LDAP test is broken as in #86305.

dasJ commented:
The code looks good to me, but I haven't tested it (as I don't use most of the stuff), which is why I'm commenting instead of approving.

Thanks for spotting my mistake with the groups, looks good now.

flokli commented:
Maybe ping @xastor and @ajs124 on the samba changes, as there are no VM tests testing this currently.

ajs124 commented:
I say what @dasJ says.

dasJ commented:
@ajs124 As a general rule of thumb or only in this particular case?

flokli commented:
squashed the fixup commit and rebased on master.

andir commented on nixos/modules/config/nsswitch.nix:
```diff
@@ -95,7 +95,9 @@ with lib;
+        # If disabling nscd is really necessary, it's still possible to opt out
+        # by forcing config.system.nssModules to [].
         assertion = config.system.nssModules.path != "" -> config.services.nscd.enable;
         message = "Loading NSS modules from path ${config.system.nssModules.path} requires nscd being enabled.";
```
IMHO that message should also tell you which option triggers this not just the value of it. Mentioning both option names (systemd.nssModules.path and services.nscd.enable) is probably desireable here. You shouldn't need a nixpkgs checkout to figure out what the condition to the message is. Often the code is much cleaner than the messages.

flokli replied:
Updated.

flokli replied:
The later version didn't interpolate `config.system.nssModules.path` anymore, so it was harder to figure out what nss modules were configured. I'll change this to the following:

> "Loading NSS modules from system.nssModules.path (${config.system.nssModules.path}) requires services.nscd.enable being set to true."

andir commented on nixos/modules/system/boot/systemd.nix:
```diff
@@ -829,16 +829,9 @@ in
-    # check for nscd being enabled before adding to nssModules.
-    system.nssModules = optional config.services.nscd.enable systemd.out;
-    system.nssDatabases = mkIf config.services.nscd.enable {
+    # enable the systemd-provided nss modules
```
this states the obvious, what make you think this comment is good? Maybe we can improve on it if it is important. We went from a very large (somtimes vague) comment to a very obvious statement.

flokli replied:
The comment in `nsswitch.nix` is essentially what explains why we need to pass nss modules via nscd, and why it's a bad idea to not do it. I could just drop this comment here, as it just writes what's written again in code immediately afterwards.

flokli replied:
I dropped the comment.

andir commented on nixos/modules/services/misc/sssd.nix:
```diff
@@ -42,11 +42,6 @@ in {
   };
   config = mkMerge [
     (mkIf cfg.enable {
-      assertions = singleton {
```
This is probably not the right move here. We are removing the option because somewhere else we have a `foo -> nscd.enabled` condition. if `foo` isn't the falsey value you could still enable `sssd` while `nscd` would be allowed to be disabled. Is that really desired?

flokli replied:
I don't understand this commend. `sssd` had its own additional assertion to complain when nscd was not enabled - but without nscd, sssd is as broken as other NSS modules.

`services.sssd.enable` sets `system.nssModules` (as any other module providing NSS modules), and we have a more global assertion complaining about this in `nsswitch.nix`.

which is now h

flokli replied:
I further improved the error message in `nsswitch.nix`, to also show the generated path of nssModules, so this should be covered.

andir reviewed (changes requested):
Does look reasonable to me.

Do we have some tests or can you write some integration tests with beaker? ;)

flokli committed:
nixos/sssd remove redundant condition

This is all inside a global cfg.enable conditional, so we don't need to
check here again.

flokli committed:
nixos/nsswitch: improve error message

Show the config option triggering the assertion, so people don't
necessary lookup the nixpkgs source code.

flokli commented:
Improved the error message as described, and rebased on top of master another time. PTAL.

flokli commented:
Okay, let's cook this in unstable.

orivej commented:
The current error message is not very helpful, I still had to look at the source code, look up the commit and this PR to understand what is going on:

    Loading NSS modules from system.nssModules (/nix/store/…), requires services.nscd.enable being set to true.

I'd suggest a more actionable error message:

    NSS modules from system.nssModules (/nix/store/…) will not be loaded without nscd, please set system.nssModules = lib.mkForce [ ]; or services.nscd.enable = true;

(I guess that those who have already disabled nscd would rather keep it disabled and do not need this assertion at all. However, the new users might like to be warned when they disable nscd for the first time…)

flokli commented:
@orivej the problem was that people disabling `nscd` didn't really notice they had their system with a dysfunctional NSS system.

Most code adding to nss modules only did this "optionally" ( when nscd was enabled (and by doing this, never triggering the assertion). However, there's a lot of places in NixOS where we rely on a functional NSS system, so alerting in these cases by actually triggering the assertion is preferrable.

The suggested error message isn't that clear either on what's preferred now. If we want to avoid people having to read the source, we probably need to be a bit more verbose. What about this? 

> NSS modules from system.nssModules (/nix/store/…) will not be loaded unless nscd is enabled. Disabling nscd is stongly discouraged, as a dysfunctional NSS system breaks features already used in a base NixOS system, such as systemd dynamic user lookups or container hostname resolution. If you really know what you're doing, you can override this by setting `system.nssModules = lib.mkForce [ ];`.

Can you elaborate on usecases to disable nscd? When would one explicitly want to opt in to have broken NSS lookups?

orivej commented:
I have disabled nscd because I do not want to erase the boundary between Linux network namespaces when it comes to DNS — otherwise the requests from nscd-aware applications originate from the nscd namespace rather than the application namespace. (Here are the relevant bits of my config: https://github.com/NixOS/nixpkgs/issues/52411#issuecomment-450634952.) I know that this disables some systemd-based name resolution and I'm OK with that since I never encounter such names.

I'd suggest this edit:

> NSS modules from system.nssModules (/nix/store/…) will not be loaded unless nscd is enabled. Disabling nscd breaks systemd dynamic user lookup and container hostname resolution. To enable nscd, do not set services.nscd.enable = false; to proceed without nscd, set system.nssModules = lib.mkForce [ ].

flokli commented:
@orivej thanks for the pointer! That's an interesting usecase. I wonder if you could just make nscd unavailable from these namespaced units, via [`BindPaths`](https://www.freedesktop.org/software/systemd/man/systemd.exec.html#BindPaths=), or whether there's a way to restrict this to only not resolve hosts via nscd from that namespace.

I don't remember exactly how the breakages looked (need to check), but "systemd dynamic user lookup and container hostname resolution" are only examples on a basic installation - with disabled nscd, all custom nss modules will break.
